lib/inputs/dynamodb-data.js

```diff
@@ -0,0 +1,156 @@
+const logger = require('../logger');
+const timestamp = require('../timestamp');
+const safeParseInt = str => {
+  const int = parseInt(str, 10);
+  return isNaN(int) ? str : int;
+};
+const OVERALL_DOWNLOAD = 'DOWNLOAD';
+
+/**
+ * Cache redirect-data and segment-bytes-downloaded-data into DynamoDB. ONLY
+ * after both of those are present (there's a race condition!) do we throw the
+ * redirect-data back onto kinesis to generate the real downloads/impressions.
+ */
+module.exports = class DynamodbData {
+  constructor(records) {
+    this._records = (records || []).filter(rec => this.check(rec));
+
+    // organize into payloads and segments, grouping by key
+    this.payloads = {};
+    this.segments = {};
+    this._records.forEach(raw => {
+      const [key, rec] = this.encodeKey(raw);
+      if (rec.type === 'antebytes' || rec.type === 'antebytespreview') {
+        this.payloads[key] = rec;
+      } else if (this.segments[key]) {
+        this.segments[key].push(this.encodeSegment(rec));
+      } else {
+        this.segments[key] = [this.encodeSegment(rec)];
+      }
+    });
+  }
+
+  check(record) {
+    return ['antebytes', 'antebytespreview', 'bytes', 'segmentbytes'].includes(record.type);
+  }
+
+  encodeKey({ listenerEpisode, digest, ...record }) {
+    return [`${listenerEpisode}.${digest}`, record];
+  }
+
+  decodeKey(key, record) {
+    const index = key.indexOf('.');
+    const listenerEpisode = key.slice(0, index);
+    const digest = key.slice(index + 1);
+    return { listenerEpisode, digest, ...record };
+  }
+
+  encodeSegment(rec) {
+    const epochMs = (rec.timestamp || Date.now()).toString();
+    if (rec.type === 'segmentbytes') {
+      return `${epochMs}.${rec.segment}`;
+    } else {
+      return epochMs;
+    }
+  }
+
+  decodeSegment(str) {
+    const parts = str.split('.');
+    const epoch = safeParseInt(parts[0]);
+    const segment = safeParseInt(parts[1]) || OVERALL_DOWNLOAD;
+    return [timestamp.toDateString(epoch), epoch, segment];
+  }
+
+  async insert() {
+    const allKeys = Object.keys(this.payloads).concat(Object.keys(this.segments));
+    const uniqueKeys = allKeys.filter((v, i, a) => a.indexOf(v) === i);
+
+    if (uniqueKeys.length == 0) {
+      return Promise.resolve([]);
+    } else {
+      const updates = uniqueKeys.map(k => [k, this.payloads[k], this.segments[k]]);
+      const result = await dynamo.updateAll(updates);
+      const output = [{ dest: 'dynamodb', count: uniqueKeys.length }];
+
+      // send "new" segments onwards to kinesis
+      const formatted = result.success.map(r => this.format(r)).flat();
+      if (formatted.length) {
+        const count = await kinesis.put(formatted);
+        output.push({ count, dest: `kinesis:${kinesis.stream()}` });
+      }
+
+      // just throw an error if anything failed. successful records have already
+      // been recorded in DDB, so retrying them won't cause duplicates.
+      if (result.failures.length > 0) {
```

Comment:
This really helps with error handling.  If something _did_ go wrong for any record (we do see DDB throttling us sometimes) ... the `segments` field in DDB makes sure we won't re-emit duplicate kinesis records for anything that succeeded this time around.

So we can safely retry the entire batch!

Reply:
Hmm... revisiting my assumptions here... so yes, this handles any DDB failures.  Since `dynamo.updateAll` catches all errors and returns them in the `failures` key.  And the `segments` that succeeded will be locked on the retry.

But you're right that in the case where `kinesis.put` failed, we'd throw an error + retry this kinesis batch, but DDB would have those segments "locked" the next go around.  I'm going to open a ticket for that ... it seems like the best path may be converting `kinesis.put` to the CloudWatch subscription `console.logs`.  Because those can't really fail, and if they're throttled by Kinesis, they get retried.

As for the case where buggy code causes an error between `dynamo.updateAll` and here ... maybe just have to make sure the code handles any bad data?  Because there's not a good way to catch and "unlock" the DDB segments afterwards.

Reply:
#76